### PR TITLE
Zadanie 3 - Róża Mazurek

### PR DESCRIPTION
**Pull request by rozamazurek**

### CODE REVIEW

**Review by MrChazar — APPROVED** [review]
Uwagi:
- świetnie przeprowadzone badania 
Bardzo dobra praca

**Review by Hanzoze — COMMENTED** [review]
Bardzo dobrze napisana praca. Wyniki dla poszczególnych etapów, wybrano kilka metryk i wizualizacja dla każdej na wykresach, analiza wpływu każdego z parametrów na sieć, mi się wszystko podoba

**Review by ozio1212 — APPROVED** [review]
Super poprowadzone badania, wszystkie wykresy też bardzo fajne. Ciężko się do czegokolwiek przyczepić tak szczerze.

**Comment by GregW04** [comment]
@rozamazurek
Dzięki za kod oto moje review:

- Test loss jest źle liczony (zawyżone dzielenie). CrossEntropyLoss domyślnie zwraca średnią w batchu. Ty sumujesz te średnie po batchach i dzielisz przez liczbę przykładów w całym teście (len(test_loader.dataset)), co daje wartości rzędu 0.001x. albo ustaw reduction="sum" i dziel przez liczbę przykładów, albo (prościej) dziel przez liczbę batchy. Nie jest to duży błąd, lecz trochę zaburza test_loss. Zobacz moje metryki po naprawie: 
    * Epoch 1/3 | Train loss: 0.2330 | Test loss: 0.1050 | Accuracy: 96.73% | Precision: 96.74% | Recall: 96.68%
    * Epoch 2/3 | Train loss: 0.0948 | Test loss: 0.0845 | Accuracy: 97.35% | Precision: 97.39% | Recall: 97.31%
    * Epoch 3/3 | Train loss: 0.0669 | Test loss: 0.0862 | Accuracy: 97.15% | Precision: 97.12% | Recall: 97.16%
    *  Parametry eksperymentu:
    *    n_epochs: 3
    *    batch_size: 64
    *    lr: 0.001
    *    optimizer_name: adam
    *    dropout_rate: 0.5
    * 
    *  Wyniki końcowe:
    *    Train Loss: 0.0669
    *    Test Loss:  0.0862
    *    Accuracy:   97.15%
    *    Precision:  97.12%
    *    Recall:     97.16%
        * Jest tez case ziarna polecam ustawić stale, dla powtarzalności eksperymentów 
        * Poprawka kodu 
```
# CHANGED: train_losses liczymy jako średnią ze wszystkich batchy w epoce
    running_train_loss = 0.0
    for data, target in train_loader:
        ...
        running_train_loss += loss.item()
    epoch_train_loss = running_train_loss / len(train_loader)   # NEW: poprawka
    train_losses.append(epoch_train_loss)

# CHANGED: test_loss liczymy jako średnią po batchach (nie dzielimy przez dataset!)
    running_test_loss = 0.0
    with torch.no_grad(): 
        for data, target in test_loader:
            ...
            running_test_loss += loss.item()
            ...

    epoch_test_loss = running_test_loss / len(test_loader)   # NEW: poprawne uśrednianie
    test_losses.append(epoch_test_loss)

```
-   Podobny problem przy run_experiment_with_softmax()
-  Baseline model lekko zbyt dobry, ale to nie szkodzi (staramy się dać bardzo prosty)
-  Polecam jeszcze podzielic na train -val -test. Trenujesz na train, oceniasz na val oraz majstrujesz ew. a na Test patrzysz wyniki i nie dotykasz już architektur modeli - 
- Fajne podzielenie na funkcje lecz jeszcze wydzilibym funkcje pomocnicze dla train i test oddzielnie (dosyć mało czytelna taka duża funkcja)
-  Oprócz tego wizualizacje super, fajnie widać na pierwszych epokach poprawę/lub nie , Jedyne co bym dodał ostatnią epokę jeszcze. + 
- Jak masz czas odpal sobie na więcej epok i obserwuj.
- Wnioski ok! + 
- BatchNorm + 
- Rozszerzanie +
- CrossEntropy vs. Softmax + 
- metryki +
- eksperymenty z optimizerami + 
- polecam jeszcze spróbować wydłużyć sieć i pobawić się różnymi funkcjami aktywacji 
- Praca przejrzysta i fajne eksperymentowanie, dzięki. +

**Comment by GregW04** [comment]
@rozamazurek dodalem jeszcze pare nowych uwag do ostatniej wiadomości
